packages/react-storage/src/components/StorageBrowser/context/actions/listLocationsAction.ts

```diff
@@ -61,21 +61,23 @@ export const createListLocationsAction = (
     let nextNextToken: ListLocationsOutput['nextToken'] = refresh
       ? undefined
       : nextToken;
-    let remainingPageSize = pageSize;
 
     do {
-      remainingPageSize = remainingPageSize - locationsResult.length;
+      const remainingPageSize = pageSize - locationsResult.length;
 
       const output = await listLocations({
         nextToken: nextNextToken,
         pageSize: remainingPageSize,
       });
       nextNextToken = output.nextToken;
 
-      locationsResult = [...locationsResult, ...output.locations].filter(
-        ({ permission, type }) =>
-          !(type === 'OBJECT' || shouldExclude(permission, exclude))
-      );
+      locationsResult = [
+        ...locationsResult,
+        ...output.locations.filter(
```

Comment:
optimized to only filter the new results we obtain each time instead of the entire array